Documentation: add "Using SBL on UP2 board " (#2222)
projectacrn/acrn-hypervisor: doc/tutorials/using_sbl_on_up2.rst

      |                              | to flash SOS image + hypervisor/SOS boot image +  |
      |                              | SOS userland                                      |
      +------------------------------+---------------------------------------------------+

.. note::
   In this step, please build SOS and LaaG images in Clear Linux rather than Ubuntu. 

Download and install flash tool
*******************************

#. Download Intel® Platform Flash Tool Lite from `<https://github.com/projectceladon/tools/tree/master/platform_flash_tool_lite/latest/>`_.

#. For Ubuntu host, install `platformflashtoollite_5.8.9.0_linux_x86_64.deb <https://github.com/projectceladon/tools/blob/master/platform_flash_tool_lite/latest/platformflashtoollite_5.8.9.0_linux_x86_64.deb>`_ for example.

SOS and LaaG Installation
*************************

#. Connect a USB cable from the debug board to your Ubuntu host machine, 
   and run the following command to verify that its USB serial port is 
   discovered and showing under ``/dev``.

   .. code-block:: none

       ls /dev/ttyUSB*
       /dev/ttyUSB0

> #. Connect to board via ``minicom``, and use ``/dev/ttyUSB0`` for exmaple:

Reviewer: ```suggestion
#. Connect to board via ``minicom``, and use ``/dev/ttyUSB0`` for example:
```

        0|      4|   MEM |    0 |   0 |    0 |  RAW |    0 | 0x0                      
        1|      0|   MMC |    0 |   0 |    0 |  RAW |    1 | 0x0   
         
         
#. Exit and reboot to fastboot mode:

   .. code-block:: none
   
       Shell> exit
      
       …
       
        40E0 |  175118 ms |     158 ms | Kernel setup
        40F0 |  175144 ms |      26 ms | FSP ReadyToBoot/EndOfFirmware notify
        4100 |  175144 ms |       0 ms | TPM IndicateReadyToBoot
       ------+------------+------------+----------------------------------

       Starting MB Kernel ...

        abl cmd 00: console=ttyS0,115200
        abl cmd 00 length: 20
        abl cmd 01: fw_boottime=175922
        abl cmd 01 length: 18
       boot target: 1
       target=1
       Enter fastboot mode ...
       Start Send HECI Message: EndOfPost
       HECI sec_mode 00000000
       GetSeCMode successful
       GEN_END_OF_POST size is 4
       uefi_call_wrapper(SendwACK) =  0
       Group    =000000FF
       Command  =0000000C
       IsRespone=00000001
       Result   =00000000
       RequestedActions   =00000000
       USB for fastboot transport layer selected

> #. When UP2 board is fastboot mode, you should be able 

Reviewer: ```suggestion
#. When UP2 board is in fastboot mode, you should be able 
```

.. _using-sbl-up2:

Using SBL on UP2 Borad
######################

> This document builds on the :ref:`getting-started-up2`, and explains how to use

Reviewer 2: Put the description of SBL (lines 38-41) here.

SBL instead of UEFI to boot UP2 board.

.. image:: images/sbl_boot_flow_UP2.png
   :align: center
   
We show a verified Boot Sequence with SBL on an Intel® Architecture platform UP2, 
and the boot process proceeds as follows:

#. SBL verifies and boots the ACRN hypervisor and Service OS kernel

> #. Service OS kernel verifies and loads ACRN Device Model and vSBL through ``dm-verity``

Reviewer 2: should this be ``dm-verify``?

* USB to TTL serial cable (PL2303TA for example) for debug UART output
* Micro USB OTG cable for flashing
* Linux host
* Internet access

.. image:: images/up2_sbl_connections.png
   :align: center

Build SBL
*********

Slim Bootloader is an open-source boot firmware solution, 
built from the ground up to be secure, lightweight, and highly 
optimized while leveraging robust tools and libraries from 
the EDK II framework.

Follow the steps of `Building <https://slimbootloader.github.io/supported-hardware/up2.html#building>`_
and `Stitching <https://slimbootloader.github.io/supported-hardware/up2.html#stitching>`_ 
from `<https://slimbootloader.github.io/supported-hardware/up2.html>`_ to generate the 
BIOS binary file ``<SBL_IFWI_IMAGE>``, which is the new IFWI image with SBL in BIOS region.

Flash SBL on the UP2
********************

#. Download the appropriate BIOS update for `UP2 Board <https://downloads.up-community.org/download/up-board-uefi-bios-upc1dm15/>`_.
#. Put the empty USB flash drive in your PC and format it as FAT32.
#. Decompress the BIOS zip file into the formatted drive. 
#. Attach the USB disk and keyboard to the board and power it on.

> #. During boot, press :kbd:`F7` on the usb keyboard to enter the UEFI BIOS boot menu.

Reviewer 2: ```suggestion
#. During boot, press :kbd:`F7` on the keyboard to enter the UEFI BIOS boot menu.
```

> #. Navigate through the following menus and select ``Build-in EFI shell``.

Reviewer 2: I think this is ``Built-in EFI shell``

#. Please take note of which filesystem number ``fs*`` your USB drive is mapped to.

> #. Switch to that filesystem, e.g. ``fs1:``.

Reviewer 2: ```suggestion
#. Switch to that filesystem, e.g. ``fs1:``.  (Don't forget the colon.)
```

#. Navigate to the path where you decompressed the update (the ``cd`` and ``ls`` commands are available here, as if in an Unix shell).
  
   .. code-block:: none
  
      Fpt_3.1.50.2222.efi -f <SBL_IFWI_IMAGE> -y

> .. note::

Reviewer 2: Too late in the instructions, so delete this note here.

   and the programming software version is ``Fpt_3.1.50.2222.efi``.

Build ACRN for UP2
******************

In Clear Linux, build out the SOS and LaaG image with these two files:

* create-up2-images.sh
* uos.json

An example of the configuration file ``uos.json``:

.. code-block:: none

   {
       "DestinationType" : "virtual",
       "PartitionLayout" : [ { "disk" : "clearlinux.img", "partition" : 1, "size" : "100M", "type" : "EFI" },
                             { "disk" : "clearlinux.img", "partition" : 2, "size" : "10G", "type" : "linux" } ],
       "FilesystemTypes" : [ { "disk" : "clearlinux.img", "partition" : 1, "type" : "ext2" },
                             { "disk" : "clearlinux.img", "partition" : 2, "type" : "ext4" } ],
       "PartitionMountPoints" : [ { "disk" : "clearlinux.img", "partition" : 1, "mount" : "/boot" },
           		          { "disk" : "clearlinux.img", "partition" : 2, "mount" : "/" } ],
       "Version": 26880,
       "Bundles": ["kernel-iot-lts2018", "openssh-server", "software-defined-cockpit", "os-core", "os-core-update"]
   }

.. note::
   To generate the image with a specified version, please modify 

>    the "Version" argument, and we can set ``"Version": 26000`` instead 

Reviewer 2: ```suggestion
   the "Version" argument, ``"Version": 26000`` instead 
```

   of ``"Version": 26880`` for example.

Clone the source code of ``acrn-hypervisor`` and build SOS and LaaG image:

.. code-block:: none
   
   cd ~
   git clone https://github.com/projectacrn/acrn-hypervisor

>    sudo ./acrn-hypervisor/devicemodel/samples/up2/create-up2-images.sh --images-type all --clearlinux-version 26880 --laag-json uos.json --acrn-code-path ~/acrn-hypervisor/

Reviewer 2: Breaking this line will display better:
```suggestion
   sudo ./acrn-hypervisor/devicemodel/samples/up2/create-up2-images.sh --images-type all \
      --clearlinux-version 26880 --laag-json uos.json --acrn-code-path ~/acrn-hypervisor/
```

      |                              | contains the Device Models implementation and     |
      |                              | SOS user space.                                   |
      +------------------------------+---------------------------------------------------+
      | partition_desc.bin           | This is the binary image for GPT partitions       |
      +------------------------------+---------------------------------------------------+
      | up2_laag.img                 | This is the root filesystem image for the SOS.    |
      |                              | It has an integrated kernel and userspace.        |
      +------------------------------+---------------------------------------------------+
      | flash_LaaG.json              | Configuration file for Intel® Platform Flash Tool |
      |                              | to flash SOS image + hypervisor/SOS boot image +  |
      |                              | SOS userland                                      |
      +------------------------------+---------------------------------------------------+

.. note::
   In this step, please build SOS and LaaG images in Clear Linux rather than Ubuntu. 

> Download and install flash tool

Reviewer: I was looking for a webpage that can be used as a general intro to the Intel Platform Flash Tool Lite but the best I found was this: https://github.com/projectceladon/tools. Can you add a link to this as, despite not being a proper page, it has the merit of detailing which host OS are supported (useful if someone does not want to use Ubuntu for example).

Author: The following download link of "Intel® Platform Flash Tool Lite" is https://github.com/projectceladon/tools/tree/master/platform_flash_tool_lite/latest.

* USB to TTL serial cable (PL2303TA for example) for debug uart output
* Micro USB OTG cable for flashing
* Linux host
* Internet access

.. image:: images/up2_sbl_connections.png
   :align: center

Build SBL
*********

Slim Bootloader is an open-source boot firmware solution, 
built from the ground up to be secure, lightweight, and highly 
optimized while leveraging robust tools and libraries from 
the EDK II framework.

Follow the steps of `Building <https://slimbootloader.github.io/supported-hardware/up2.html#building>`_
and `Stitching <https://slimbootloader.github.io/supported-hardware/up2.html#stitching>`_ 
from `<https://slimbootloader.github.io/supported-hardware/up2.html>`_ to generate the 
BIOS binary file ``<SBL_IFWI_IMAGE>``, which is the new IFWI image with SBL in BIOS region.

Flash UEFI to SBL
*****************

#. Download the appropriate bios update for `UP2 Board <https://downloads.up-community.org/download/up-board-uefi-bios-upc1dm15/>`_.
#. Put the empty USB flash drive in your pc and format it to FAT32.
#. Decompress the bios zip file into the formatted drive. 
#. Attach the USB disk to the board and a usb keyboard and power it on.

> #. During boot, hit F7 on the usb keyboard to enter the UEFI BIOS boot menu.

Reviewer 2: ```suggestion
#. During boot, press :kbd:`F7` on the keyboard to enter the UEFI BIOS boot menu.
```

#. Navigate through the following menus and select ``Build-in EFI shell``.
#. Please take note of which filesystem number ``fs*`` your USB drive is mapped to.

> #. Switch to that filesystem.

Reviewer 2: It's a common problem to leave off the colon on this command too, so you could add "note the trailing colon on this command", as a reminder.

> #. Navigate to the path where you decompressed the update (the cd and ls commands are available here, as if in an Unix shell).

Reviewer 2: ```suggestion
#. Navigate to the path where you decompressed the update (the ``cd`` and ``ls`` commands
   are available here, as if in a Unix shell), and then run the command:
```
